docs/csharp/advanced-topics/performance/index.md

```diff
@@ -0,0 +1,106 @@
+
+You can also use *ref assignment* to have a variable refer to another variable. A typical assignment copies the *value* of the right hand side to the variable on the left hand side of the assignment. A *ref assignment* copies the memory location of the variable on the right hand side to the variable on the left hand side. The `ref` now refers to the original variable:
+
+:::code language="csharp" source="./snippets/ref-safety/Program.cs" id="RefAssignment":::
+
+When you *assign* a variable, you change its *value*. When you *ref assign* a variable, you change what it refers to.
+
+You can work directly with the storage for values using `ref` variables, pass by reference, and ref assignment. Scope rules enforced by the compiler ensure safety when working directly with storage.
+
+## Ref safe to escape scope
+
+C# includes rules for `ref` expressions to ensure that a `ref` expression can't be accessed where the storage it refers to is no longer valid. Consider the following example:
+
+```csharp
+public ref int CantEscape()
+{
+    int index = 42;
+    return ref index; // Error: index's ref safe to escape scope is the body of CantEscape
+}
+```
+
+The compiler reports an error because you can't return a reference to a local variable from a method. The caller can't access the storage being referred to. The *ref safe to escape scope* defines the scope in which a `ref` expression is safe to access or modify. The following table lists the *ref safe to escape scopes* for variable types. `ref` fields can't be declared in a `class` or a non-ref `struct`, so those those rows aren't in the table:
+
+| Declaration                 | *ref safe to escape scope*    |
+|-----------------------------|-------------------------------|
+| non-ref local               | block where local is declared |
+| non-ref parameter           | current method                |
+| `ref`, `in` parameter       | calling method                |
+| `out` parameter             | current method                |
+| `class` field               | calling method                |
+| non-ref `struct` field      | current method                |
+| `ref` field of `ref struct` | calling method                |
+
+A variable can be `ref` returned if its *ref safe to escape scope* is the calling method. If its *ref safe to escape scope* is the current method or a block, `ref` return is disallowed. The following snippet shows two examples. A member field can be accessed from the scope calling a method, so a class or struct field's *ref safe to escape scope* is the entire method. The *ref safe to escape scope* for a parameter with the `ref`, or `in` modifiers is the entire method. Both can be `ref` returned from a member method:
```

Comment:
>is the entire method [](http://example.com/codeflow?start=451&length=20)

The table above looks correct. The comment here that the ref safe to escape scope for a `ref` or `in` parameter is the entire method might be confusing though since the table says _calling method_ and the sentence following this one says the `ref` can be returned from the method. Consider using _calling method_ here instead of _entire method_ to reduce the chance of confusion.